lib/signalfx.js

```diff
@@ -0,0 +1,334 @@
+      properties: properties || []
+    };
+  };
+
+
+// ============================================================================================
+//                          Client Signal Fx class - base API client
+// ============================================================================================
+  /**
+   * SignalFx API client.
+   * This class presents a programmatic interface to SignalFx's metadata and
+   * ingest APIs. At the time being, only ingest is supported; more will come
+   * later.
+   * @param apiToken
+   * @constructor
+   */
+  function SignalFxClient(apiToken) {
+    BaseSignalFx.apply(this, arguments);
+
+    this.HEADER_API_TOKEN_KEY = 'X-SF-Token';
+    this.HEADER_USER_AGENT_KEY = 'User-Agent';
+    this.HEADER_CONTENT_TYPE = 'Content-Type';
+
+    this.INGEST_ENDPOINT_SUFFIX = 'v2/datapoint';
+    this.API_ENDPOINT_SUFFIX = 'v1/event';
+
+    this.queue = [];
+    this.asyncRunning = false;
+  }
+
+  SignalFxClient.prototype = Object.create(BaseSignalFx.prototype);
+  SignalFxClient.prototype.constructor = BaseSignalFx;
+
+  /**
+   * Send the given metrics to SignalFx.
+   *
+   * @param cumulativeCounters (list): a list of dictionaries representing the cumulative counters to report.
+   * @param gauges (list): a list of dictionaries representing the gauges to report.
+   * @param counters (list): a list of dictionaries representing the counters to report.
+   */
+  SignalFxClient.prototype.send = function (cumulativeCounters, gauges, counters) {
+    var data = BaseSignalFx.prototype.send.apply(this, arguments);
+
+    if (data === null || data === undefined) {
+      return;
+    }
+
+    var _this = this;
+    for (var metricType in data) {
+      if (data.hasOwnProperty(metricType)) {
+        var dataPoints = data[metricType];
+        if (dataPoints !== null && dataPoints !== undefined) {
+          if (!Array.isArray(dataPoints)) {
+            throw new TypeError('DataPoints not of type list: ' + dataPoints);
+          }
+          dataPoints.forEach(function (dataPoint) {
+            _this.addToQueue(metricType, dataPoint);
+          });
+        }
+      }
+    }
+
+    this.startAsyncSend();
+  };
+
+  SignalFxClient.prototype.addToQueue = function (metricType, datapoint) {
+    throw new Error('Subclasses should implement this!');
+  };
+
+  SignalFxClient.prototype.batchData = function (datapointsList) {
+    throw new Error('Subclasses should implement this!');
+  };
+
+  /**
+   * Send an event to SignalFx
+   *
+   * @param eventType (string): the event type (name of the event time series).
+   * @param dimensions  (dict): a map of event dimensions.
+   * @param properties  (dict): a map of extra properties on that event.
+   */
+  SignalFxClient.prototype.sendEvent = function (eventType, dimensions, properties) {
+    var data = BaseSignalFx.prototype.sendEvent.apply(this, arguments);
+    if (data === null || data === undefined) {
+      return null;
```

Comment:
This returns null when data isn't defined but .send returns undefined, is this on purpose or is it an inconsistency?